Brokerages/Bitfinex/BitfinexBrokerage.cs

```diff
@@ -284,31 +284,53 @@ public override IEnumerable<BaseData> GetHistory(Data.HistoryRequest request)
                 yield break;
             }
 
+            if (request.StartTimeUtc >= request.EndTimeUtc)
+            {
+                OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Warning, "InvalidDateRange",
+                    "The history request start date must precede the end date, no history returned"));
+                yield break;
+            }
+
             string resolution = ConvertResolution(request.Resolution);
-            long resolutionInMS = (long)request.Resolution.ToTimeSpan().TotalMilliseconds;
+            long resolutionInMsec = (long)request.Resolution.ToTimeSpan().TotalMilliseconds;
             string symbol = _symbolMapper.GetBrokerageSymbol(request.Symbol);
-            long startMTS = (long)Time.DateTimeToUnixTimeStamp(request.StartTimeUtc) * 1000;
-            long endMTS = (long)Time.DateTimeToUnixTimeStamp(request.EndTimeUtc) * 1000;
+            long startMsec = (long)Time.DateTimeToUnixTimeStamp(request.StartTimeUtc) * 1000;
+            long endMsec = (long)Time.DateTimeToUnixTimeStamp(request.EndTimeUtc) * 1000;
             string endpoint = $"v2/candles/trade:{resolution}:t{symbol}/hist?limit=1000&sort=1";
+            var period = request.Resolution.ToTimeSpan();
 
-            while ((endMTS - startMTS) > resolutionInMS)
```

Comment:
Would simply updating this to `>=` also permit this case to work?

Reply:
@mchandschuh ```>=``` would also work just fine. 

I now remember that when I was first editing a file I was also substracting a millisecond from end time :
```long endMsec = (long)Time.DateTimeToUnixTimeStamp(request.EndTimeUtc) * 1000 - 1; ```
and therefore do-while was then necessary. 

A reason for this is that Bitfinex makes trade bar's open time a timestamp. And if requesting data for the interval where the beginning and end are rounded down - as Lean I believe is doing all over the place - it will give us also return bar that starts at end time. Example. Lets say we requst data from 13:05:00 to 13:07:00 --> this will return three bars: 
13:05:00 to 13:06:00
13:06:00 to 13:07:00
13:07:00 to 13:08:00

I thought it does no harm an extra bar but if to understand the end time as an upper cap - no data is to be received after - we might want to make a correction for this and subtract a millisecond... ?

Reply:
RE: _example_:

I think that's a bug if we're returning data ahead of the end time. LEAN should _always_ be using end times for determining time synchronization issues, such as determining what bars fall within a requested time span. An even more egregious example would be when using daily data. If we request from 2019.09.17 to 2019.09.18, I would expect to get a single daily bar for the date of the 17th and I would **not** expect to also receive the bar for the 18th.

RE: _subtract a millisecond_:
Individual implementations can do what they need to do in order to fulfill their responsibilities. Subtracting a millisecond _may_ work, but another way to look at it is to subtract the resolution for bitfinex. If bitfinex uses start times than we can project the user's request to reflect that. For example, LEAN _should_ be using (start, end] semantics, meaning, we include all day from start to end _including_ the start and _excluding_ the end. If bitfinex's api looks at it as (start, end-resolution], then perhaps we just modify the end time of the request to shift it to a bar start time. I'm always a little hesitant to subtract a millisecond here or a tick there as it appears to be more of a bandaid than a projection of data to fit the api's conventions.

Reply:
@mchandschuh agree that it's reasonble to drop the last bar. please kindly review the commit below that serves to solve an issue. things have been changed here:

1) Additional warning was added to ensure end time is a full round of history request resolution. Idea is that bitfinex API may provide only historical bars and to avoid any ambiguity the user must be precise about timing borders. ?
2) Additional Logic to drop the last candle that start at end TimeKeeper. Solution that would modify end time by substracting the resolution amount of milliseconds will not work for the history request of 1-bar-long, where the start and time time will result in to be the same - exchange will return an error.
3) & 4) To test the things work as expected Bitfinex History provider tests have been little modified. To support the rounding agreement. And to provide the valid test case of 1 bar long historical request:
```new TestCaseData(Symbol, Resolution.Minute, Time.OneMinute, false, false),```
to ensuse this does return the data